pkg/quota/image/usage.go

```diff
@@ -10,139 +13,400 @@ import (
+						continue
+					}
 
-			for _, layer := range img.DockerImageLayers {
-				if _, exists := processedLayers[layer.Name]; exists {
+				} else {
+					tag = ref.Tag
+					if len(tag) == 0 {
+						tag = "latest"
+					}
+					ist, err2 := c.osClient.ImageStreamTags(ref.Namespace).Get(ref.Name, tag)
+					if err2 != nil {
+						glog.Errorf("Failed to get image stream tag %s/%s:%s: %v", ref.Namespace, ref.Name, tag, err2)
+						continue
+					}
+					img = &ist.Image
+					c.imageCache[ref.ID] = img
+				}
+
+			} else {
+				img, err = c.getImage(ref.ID)
+				if err != nil {
+					glog.Errorf("Failed to get image %s: %v", ref.ID, err)
 					continue
 				}
-				isSize.Add(*resource.NewQuantity(layer.Size, resource.BinarySI))
-				processedLayers[layer.Name] = sets.Empty{}
 			}
 		}
+
+		if processedImages.Has(img.Name) {
+			glog.V(4).Infof("Skipping image %q - already processed", img.Name)
+			continue
+		}
+		processedImages.Insert(img.Name)
+
+		if value, ok := img.Annotations[imageapi.ManagedByOpenShiftAnnotation]; !ok || value != "true" {
+			glog.V(4).Infof("Image %q with DockerImageReference %q belongs to an external registry - skipping", img.Name, img.DockerImageReference)
+			continue
+		}
+
+		cacheInternalRegistryName(img.DockerImageReference)
+
+		if err := handler(tag, img.DockerImageReference, img); err != nil {
+			return err
+		}
 	}
 
-	for _, layer := range image.DockerImageLayers {
-		if _, exists := processedLayers[layer.Name]; exists {
-			sizeIncrement.Sub(*resource.NewQuantity(layer.Size, resource.BinarySI))
-			if sizeIncrement.Value() <= 0 {
-				sizeIncrement.Set(0)
-				break
-			}
+	return nil
+}
+
+// getImageReferenceForObjectReference returns corresponding docker image reference for the given object
+// reference representing either an image stream image or image stream tag or docker image.
+func (c *GenericImageStreamUsageComputer) getImageReferenceForObjectReference(
+	namespace string,
+	objRef *kapi.ObjectReference,
+) (imageapi.DockerImageReference, error) {
+	switch objRef.Kind {
+	case "ImageStreamImage":
+		nameParts := strings.Split(objRef.Name, "@")
+		if len(nameParts) != 2 {
+			return imageapi.DockerImageReference{}, fmt.Errorf("failed to parse name of imageStreamImage %q", objRef.Name)
+		}
+		res, err := imageapi.ParseDockerImageReference(objRef.Name)
+		if err != nil {
+			return imageapi.DockerImageReference{}, err
+		}
+		if res.Namespace == "" {
+			res.Namespace = objRef.Namespace
+		}
+		if res.Namespace == "" {
+			res.Namespace = namespace
+		}
+		return res, nil
+
+	case "ImageStreamTag":
+		// This is really fishy. An admission check can be easily worked around by setting a tag reference
+		// to an ImageStreamTag with no or small image and then tagging a large image to the source tag.
+		// TODO: Shall we refuse an ImageStreamTag set in the spec if the quota is set?
+		nameParts := strings.Split(objRef.Name, ":")
```

Comment:
The one I found wasn't exported and I lacked enough courage to do so. I'll do better this time :).